build/metadata_utilities.py

```diff
@@ -37,46 +121,108 @@ def test_merge_dict_second_is_empty():
 
 def test_merge_dict_replace_in_dict_and_list():
     a        = {'a':1, 'b':{'b1':2, 'b2':3}, 'c':['x', 'y', 'z']}
     b        = {'b':{'b3':7, 'b1':8}, 'c':{0:'r', 1:'s', 2:'t'}}
     expected = {'a':1, 'b':{'b1':8, 'b2':3, 'b3':7}, 'c':['r', 's', 't']}
-    _do_the_test(a,b, expected)
+    _do_the_test_merge_dicts(a,b, expected)
 
 def test_merge_dict_with_regex():
     a        = {'aaa':{}, 'aaaa':{'x':98}, 'aaaaa':{}, 'bbb':{'bbb1':2, 'bbb2':3}, 'ccc':['x', 'y', 'z']}
     b        = {'a+':{'z':99}}
     expected = {'aaa':{'z':99}, 'aaaa':{'x':98, 'z':99}, 'aaaaa':{'z':99}, 'bbb':{'bbb1':2, 'bbb2':3}, 'ccc':['x', 'y', 'z']}
-    _do_the_test(a,b, expected)
+    _do_the_test_merge_dicts(a,b, expected)
+
+def _do_the_test_add_all_metadata(funtions, expected):
+    actual = copy.deepcopy(funtions)
+    actual = add_all_metadata(actual)
+    assert expected == actual, "\nfunctions = {0}\nexpected = {1}\nactual = {2}".format(pp.pformat(funtions), pp.pformat(expected), pp.pformat(actual))
+
+def test_add_all_metadata_simple():
+    functions = {
+        'close': {
+            'codegen_method': 'public',
+            'returns': 'ViStatus',
+            'parameters': [
+                {
+                    'direction': 'in',
+                    'enum': None,
+                    'name': 'vi',
+                    'type': 'ViSession',
+                    'documentation': {
+                        'description': 'Identifies a particular instrument session.',
+                    },
+                },
+            ],
+            'documentation': {
+                'description': 'Closes the specified session and deallocates resources that it reserved.',
+            },
+        },
+    }
+    expected = {
+        'close': {
+            'codegen_method': 'public',
+            'documentation': {
+                'description': 'Closes the specified session and deallocates resources that it reserved.'
+            },
+            'parameters': [
+                {
+                    'ctypes_type': 'ViSession_ctype',
+                    'ctypes_variable_name': 'vi_ctype',
+                    'direction': 'in',
+                    'documentation': {
+                        'description': 'Identifies a particular instrument session.'
+                    },
+                    'enum': None,
+                    'intrinsic_type': 'int',
+                    'is_buffer': False,
+                    'name': 'vi',
+                    'python_name': 'vi',
+                    'python_type': 'ViSession',
+                    'size': {   'mechanism': 'fixed',
+                                'value': 1},
+                    'type': 'ViSession'
+                }
+            ],
+            'python_name': 'close',
+            'returns': 'ViStatus',
+            'returns_ctype': 'ViStatus_ctype',
+            'returns_python': 'ViStatus'
+        }
+    }
+
+    _do_the_test_add_all_metadata(functions, expected)
+
+
```

Comment:
This is greater, no doubt better than what we had (no tests), but definitely not thorough.
We should

* Add this file to coverage
* Open issue about writing unit tests for each function

Reply:
We are trying to get coverage data on it, but while the tests run, it doesn't seem to be reporting coverage information.